### ISSUE TITLE
Energy usage while updating display

### ISSUE THREAD

**Issue by michapr** [znanev/ATC_MiThermometer#7]
Original firmware while updating display:
![c401_partial_update](https://user-images.githubusercontent.com/11854556/100771645-f3355300-33fe-11eb-9ead-1d51b357ff14.jpg)
(take y-value/10 for mA)

Custom firmware  with `cpu_stall_wakeup_by_timer0()`
![c401_cust_fw_2](https://user-images.githubusercontent.com/11854556/100772682-37752300-3400-11eb-9e0b-e5ff9e271548.jpg)
(one time unit = 500ms)

This part comes up only while updating display with new value, any idea what can be different here?
(100ms <-> 1.5sec)

**Comment by znanev** [comment]
Ok, now I clearly see the difference and have no explanation - it seems on the second graph that the CPU is busy waiting while active (and it should be in deep sleep instead).

Just to be 100% sure that you've got the latest binary, can you please get it from the latest release:

https://github.com/znanev/ATC_MiThermometer/releases/download/0.4.0/ATC_Thermometer.bin

Cheers!

**Comment by michapr** [comment]
Here is the picture from this FW:
![c401_cust_fw_3](https://user-images.githubusercontent.com/11854556/100781723-af951600-340b-11eb-8059-e70f0ad34472.jpg)
(one unit=500ms)

**Comment by znanev** [comment]
Hm, that's really strange, indeed.

Is this measurement on startup, or is it some of the cyclical updates?

**Comment by michapr** [comment]
It is one of the cyclic updates, you see the advertising peaks every 2 seconds. 
During startup there is a longer interval.

**Comment by znanev** [comment]
I'll have to make some captures with a logic analyser from the original and then the modified firmware, but can't promise I'll find time for this during the week.

How about the startup sequence? Does the modified firmware consume much more energy than the original one?

**Comment by michapr** [comment]
I have not checked it, because there is the "ATC sequence" with MAC.
This take a long time same energy (if I remember right) - but this is only once, that's why have not checked in detail.

I have already tried to move the POWER_ON / POWER_OFF to other places, but no difference.
Do you remember about the IO_BUSY_N waiting times after each transmit operation?

**Comment by michapr** [comment]
Of course not sure (do not know this CPU in detail) - but maybe the suspend mode is not the right one here in this case?

http://wiki.telink-semi.cn/tools_and_sdk/Driver/doc/kite/html/group___g_p7.html
http://wiki.telink-semi.cn/tools_and_sdk/Driver/doc/kite/html/md__project_1__table_of__content_10__t_s_i__p_m__features.html

The _Deepsleep with SRAM Retention Mode_ seems to save more energy and can be used with timer for wakeup.
As far as I understand it is used for the sleep between the advertising intervals too (if I'm not wrong)

**Comment by znanev** [comment]
> Of course not sure (do not know this CPU in detail) ...

Neither do I - to be honest the documentation for this CPU is slacking compared to "big brands" CPUs... 

This function looks promising - `cpu_sleep_wakeup_32k_rc()`:

http://wiki.telink-semi.cn/tools_and_sdk/Driver/doc/kite/html/pm_8h.html#af8f6984e656c60eae0d02c45a573e1a9

I'll have to experiment with it, but as I said previously, my main interest was to make the EPD work .. at all :) My knowledge about this particular MCU is close to zero. But nevertheless I'm curious enough to try to at least "emulate" the original firmware regarding the EPD driving, I'm just not confident enough in my ability to delve deeper into this particular platform.

**Comment by atc1441** [comment]
I wouldnt call myself a guru on that as well, my way would also be try and error, need to get my hands on a low power meassuring method!

Another approach is to make a state mashine and use an interrupt on the busy pin. So going to sleep on update and let the epd wake the mcu.

**Comment by znanev** [comment]
@michapr Can you please try measuring the energy with the latest release:

https://github.com/znanev/ATC_MiThermometer/releases/download/0.5.0/ATC_Thermometer.bin

I changed the IDLE sleep with DEEP SLEEP this time, I'm curious to see if there are any measurable changes to the energy consumed.

@atc1441 This was exactly my train of thought too - instead of waiting to de-assert BUSY_N signal, create a pin interrupt handler for it... for which handling of course a state machine will be needed. Any volunteers for that ;)

**Comment by atc1441** [comment]
Would really love to do it, but it will take the evenings of about 1 week to be perfect and i can not spend that time right now. As the complete integration of the EPD would be on the list then as well. 
Also data logging and persistense storage^^ oh an 2 point callibration.
Sorry for OT

**Comment by znanev** [comment]
@atc1441 I was actually joking for the volunteering part ;)

No worries, I'm in the same boat - not much spare time right now, just satisfying my insatiable curiosity ...

**Comment by atc1441** [comment]
Even if its a joke it's still true :D some day someone will do it

**Comment by michapr** [comment]
@znanev 
>@michapr Can you please try measuring the energy with the latest release:
https://github.com/znanev/ATC_MiThermometer/releases/download/0.5.0/ATC_Thermometer.bin

No visible difference,... Will test more later today.

**Comment by znanev** [comment]
@michapr 
> No visible difference,... Will test more later today.

I think that I have a theory about what might be happening!

The original firmware broadcasts temperature and humidity once in about 10 minutes and battery level once in an hour. So the energy you measure on every update interval is just the one consumed by the sensor and the EPD, and does not include the BLE broadcast! In contrast, the ATC firmware is advertising measurements about every 2 seconds. So while you are measuring the EPD updates, it might just happen that the measurement also includes the radio transmission of the BLE advertisement packets! @atc1441 - can probably double-check this to be 100% sure.

So, if you want to test the energy consumption of both firmwares in a more comparable manner, I think that we should build a version with BLE advertisement disabled, or at least with a big enough interval so you can really measure just the EPD update energy consumption.

Surely, I may be totally wrong.. but at least this might be a good start !

**Comment by atc1441** [comment]
You are wrong on that. 
BLE is always broadcasting its name, in adidtion to the name it is "Advertising" the sensor data. It makes not difference if Advertising is enabled or not. Ble is sending every 3 seconds or so an name and advertising package even on stock firmware

**Comment by atc1441** [comment]
It could still be that ble is active while meassuring tho

**Comment by znanev** [comment]
Fair enough - that's why I wanted to fact-check my theory with you.

What I can do later this week is capture the EPD communication from the original firmware and from the modified one. Then compare the overall timings...

**Comment by michapr** [comment]
I have compared the refresh on display only - was waiting while display was changed and captured the picture - in both cases...

@znanev : Is for C401 still P14 still the SWS pin? I have to flash my sensor... wrong software experiment ;)
And - what is the Reset pin there, do you know it?

Thanks

**Comment by znanev** [comment]
Yep, it's the P14 pad. And don't forget to connect a 1.5k resistor between TX and RX of your USB-to-Serial adapter.

As for the reset - it is not brought to a pad - you have to use pin 18 on the chip directly, if you must.

Or just use ATCtelink.py directly - I flash mine with it all the time. But the newest one from @atc1441's repo doesn't work anymore for me - so be sure to use older version (before today), or use the one from this fork.

**Comment by michapr** [comment]
Thanks, reflashed... ;)

Want to add some useful information about original firmware...
I was thinking about the EPD ready time - this should be visible on original sensor too... And it is.
You see the advertising peaks every 2 seconds about - and the lower power usage for partial update.

So the time for update after measurement is also about 2 seconds (so should be) - but the used energy is smaller.
![c401_partial_update_long](https://user-images.githubusercontent.com/11854556/101190872-47d40a80-3659-11eb-87b4-4a3385588629.jpg)
(one time unit: 500ms)

Maybe the update procedure is other? As sample not all elements will be updated - only the different?

**Comment by znanev** [comment]
> Maybe the update procedure is other? As sample not all elements will be updated - only the different?

Surely my implementation is quite different than the one used in the original firmware. The only way to tell this is to de-compile the original code, but it is a tedious task and I couldn't configure the tools needed to do that. The other option was reverse-engineering, which I did. Based on my observations and findings, having sniffed the communication traffic between the CPU and the EPD, I'm quite confident that the communication part is quite close to the original. At least this is what I saw when re-captured said communication with logic analyser.

The original firmware performs partial update most of the time - which inevitably takes about 2 seconds every time. This is still quite a gain, because full update takes more than 4 seconds. But this is the nature of EPD displays - they need time to energise the white and black particles in order to display / clear segments.

I'll try to re-capture the original firmware's communication between the MCU and EPD with logic analyser during the weekend, so we can later compare with the custom implementation.

Meanwhile, if you are feeling adventurous, you can try flashing @atc1441 's original firmware (from his repo, not this fork) in your MHO-C401 device. Surely the display will be blank, but the BLE part should be working normally. Then do the same measurements (wait at least 10 seconds from device start, as this is the time the initializations are done and MAC address is shown). It will be interesting to see what the "baseline" is when later comparing what the firmware driving the EPD display consumes.

**Comment by michapr** [comment]
C401 with MMC3 firmware (latest from today):
**Measure temperature:**
![c401_cust_fw-mmc_measure](https://user-images.githubusercontent.com/11854556/101240706-4c4c0200-36f1-11eb-8a40-999f886bab90.jpg)

Advertising interval:
![c401_cust_fw-mmc_advertising_200ms](https://user-images.githubusercontent.com/11854556/101240760-9503bb00-36f1-11eb-87a6-209d4fe31176.jpg)
(one time unit: 200ms)

Advertising detail (typical):
![c401_cust_fw-mmc_advertising](https://user-images.githubusercontent.com/11854556/101240767-a2b94080-36f1-11eb-951a-b976ce90aa93.jpg)

For all pictures: y-value / 10 = mA

**Comment by pvvx** [comment]
The ATC_MiThermometer firmware fits completely into SRAM.
ATC_MiThermometer uses up to 45 kilobytes (codes + data + bss + stack).
The chip has 64 kilobytes of SRAM.
The current firmware is just filled with zeros of unused retention SRAM and is 60.7KB in size.
Compression is possible in the linker script, discarding unnecessary ones.
Using code in SRAM increases performance, which also affects consumption (in active-mode). 

However, storing large amounts of SRAM increases the hibernation consumption. 
Chip SRAM power saving is controlled bit by bit for 8/8/16/16 KB blocks.
The software (SDK) has it, but its options are limited in the linker scripts.

The author did not use CPU stops in wait cycles and does not disable unused CPU peripherals.
This alone gives a reduction in consumption by more than 3 times.

You can reduce consumption by working on combinations and leaving the previous algorithms up to 5 times.

**Comment by pvvx** [comment]
Original firmware (first startup)
![image](https://user-images.githubusercontent.com/12629515/101922527-3fdb1400-3bdf-11eb-8783-1314f16d52ca.png)
sleep_us(x * ms) ->
_attribute_ram_code_ void pm_WaitMs(unsigned int ms) {
cpu_stall_wakeup_by_timer0(ms*CLOCK_SYS_CLOCK_1MS);
}
![image](https://user-images.githubusercontent.com/12629515/101922840-ab24e600-3bdf-11eb-8a92-10b6d1f8aef2.png)
*(all test power 3.3V)

Turn off the ADC when not in use to measure power = - 0.4 мА

The average value only for the TLSR8251 microcircuit = 0.0130 mA = 13 μA with a simple cycle... :
![image](https://user-images.githubusercontent.com/12629515/101923632-b4628280-3be0-11eb-99fb-74a680896d5d.png)

**Comment by michapr** [comment]
@pvvx 👍 Great work, I have limited knowledge about this processor, and make experiments on sensor is a bit "dangerous" ;)
Was happy to be possible two time revive it with possible reflash :)

Can you publish your modified code anywhere ? - I really would like to test and compare it with current results.

Thanks!

**Comment by znanev** [comment]
@pvvx Thanks for your thorough analysis!

This is way over my knowledge about this particular CPU, but nevertheless your findings will surely serve as a kick start for someone more knowledgeable, who can fine-tune the code to be less power hungry.

**Comment by pvvx** [comment]
I did not check and did not alter everything. 
Old Fork with small changes laid out a long time ago.
https://github.com/pvvx/ATC_MiThermometer
PS: Basically I dig with modules (TB-03F, TB-04 and others), and not with ready-made devices ...

**Comment by michapr** [comment]
@pvvx - if I understand right you say the main difference in energy consumption  is to use the `cpu_stall_wakeup_by_timer0(ms*CLOCK_SYS_CLOCK_1MS);` ? This is already used in epd.c - the most "problematic" display update procedure.

Your graph (startup) is with EPD (MHO-C401) ?
Have you any idea what could be optimized here in this code part?

**Comment by pvvx** [comment]
The chip, when turned on, consumes a lot when it comes out of sleep mode. The delay before GPIO initialization is very important until they are all initialized. + 3 or more mA.

**Comment by pvvx** [comment]
Disable the fattest consumer - flash cache (ictag) has not yet been possible. It might not be possible to make it work only from SRAM. Until there was time...

**Comment by michapr** [comment]
The current main problem for this sensor usage is the e-ink display procedure.

As you can see in the picture at the top - the procedure take about 1.5 seconds and 3mA. Time is needed for the display - there is no way...
But the current is too high, the original firmware take some more time but less current for same thing: https://github.com/znanev/ATC_MiThermometer/issues/7#issuecomment-738892658
(take y-value/10 for mA)
It looks like the CPU sleep loop is not working effective (in "send_sequence()" while waiting for the IO_BUSY_N ), but not sure of course.

Maybe you have any hint for us?

**Comment by znanev** [comment]
> The average value only for the TLSR8251 microcircuit = 0.0130 mA = 13 μA with a simple cycle... :
> ![image](https://user-images.githubusercontent.com/12629515/101923632-b4628280-3be0-11eb-99fb-74a680896d5d.png)

@pvvx Are you sure that the consumption shown every ~2 seconds is indeed due to the EPD refresh? From what I see on the logic analyser, the display is not refreshed on uniform intervals, but only when there is a change to be shown (temperature / humidity / symbols). I'll post captures later.

In order to be sure that the display updates as frequently as possible, what I do at least, is breath out onto the sensor, then quickly waving a piece of cardboard to cool it down, then breath out again etc. @pvvx - if you can do this an re-capture the energy consumption, that will be great.

**Comment by michapr** [comment]
@znanev - as far as I understand this is for MMC03 sensor - it's about the chip power usage only.

I have modified your code a bit - updating EPD only if temp/hum changed - it save much battery as I can see. Have set advertising interval to 5 seconds. Battery measurement to large interval (some hours).

And at second sensor have optimized the EPD update only if temp. change is >0.1 degree and hum. change >2% - it is for my case ok and will save battery in the meantime, while we have not optimized the EPD update function.

In any case this is better as to connect every 5 minutes to the sensor and request from there the values...

**Comment by znanev** [comment]
I'm posting some captures from the logic analyser, made with the original firmware of MHO-C401.

1. Start-up sequence on power on.

[MHO-C401_PowerOn_Init.logicdata](https://github.com/znanev/MHO-C401/raw/master/captures/MHO-C401_PowerOn_Init.logicdata)

![image](https://user-images.githubusercontent.com/20048364/101987132-1dea9b80-3c8a-11eb-91bc-1afc6a2ca0aa.png)


Some observations:

Up until 1st second there is some glitching from inserting the USB cable of the power supply, connected to the device (A1).

- The full initialisation sequence takes 7.33 seconds (A1 to A2).
- Following command DISPLAY_REFRESH (0x12), signal BUSY_N is low for 1.57 seconds (B1 to B2 and C1 to C2). During this initialisation sequence, each DISPLAY_REFRESH command performs full update (taking 1.57 seconds). 

2. Some partial updates within one minute

[MHO_C401_PartialUpdates_within_one_minute.logicdata](https://github.com/znanev/MHO-C401/raw/master/captures/MHO_C401_PartialUpdates_within_one_minute.logicdata)

![image](https://user-images.githubusercontent.com/20048364/101987453-14623300-3c8c-11eb-976f-a0c5c977f13c.png)


Some observations:

I captured 60 seconds of date while the device was running its "normal" cycle - i.e. displaying current temperature and humidity values. As those were relatively stable with the period of one minute, there were just two display updates within this period.

- Partial update sequence (from POWER_ON - 0x04 until DISPLAY_REFRESH - 0x12) takes about 1.6 seconds
- Following command DISPLAY_REFRESH (0x12), signal BUSY_N is low for 1.25 seconds (A1 to A2 and B1 to B2). During this sequence, each DISPLAY_REFRESH command performs partial update (taking 1.25 seconds - compared to 1.57 for full update).

As you can see on the graph, there is a "silent" period of about 37 seconds, and no communication is happening between the CPU and the display during this interval. So the original firmware indeed updated the display only if there are any changes, not on regular intervals.

Later on I'll post similar captures, but with the device running the modified firmware.

**Comment by znanev** [comment]
Following up with captures from the modified (ATC) firmware, latest code (as of Release 0.6.0).

1. Start-up sequence on power on.

[MHO-C401_ATC_PowerOn_Init.logicdata](https://github.com/znanev/MHO-C401/raw/master/captures/ATC_firmware/MHO-C401_ATC_PowerOn_Init.logicdata)

![image](https://user-images.githubusercontent.com/20048364/101993009-f22ddc80-3cae-11eb-9a8c-2da2790c54b2.png)

Some observations:

The full initialisation sequence takes 6.68 seconds (A1 to A2). This is a bit faster than the original firmware (7.33), but I've set quite aggressive timings for the clock signal and these seem to work ok.

Following command DISPLAY_REFRESH (0x12), signal BUSY_N is low for 1.57 seconds (B1 to B2 and C1 to C2). During this initialisation sequence, each DISPLAY_REFRESH command performs full update (taking 1.57 seconds). This is exactly the same as in the original firmware and can be expected (i.e. this is the time waiting for the display to de-assert the BUSY_N signal).

2. Some partial updates within one minute

[MHO_C401_ATC_PartialUpdates_within_one_minute.logicdata](https://github.com/znanev/MHO-C401/raw/master/captures/ATC_firmware/MHO_C401_ATC_PartialUpdates_within_one_minute.logicdata)

![image](https://user-images.githubusercontent.com/20048364/101992572-51d6b880-3cac-11eb-8a9a-7b944f0fb8fa.png)


Some observations:

I captured 60 seconds of data while the device was running its "normal" cycle - i.e. displaying current temperature and humidity values. Although those were relatively stable (changed in fact two to three times within that minute), it is visible that the display is refreshed on every wake interval (this is what the code does and can be optimised, as @michapr mentioned above).

Partial update sequence (from POWER_ON - 0x04 until DISPLAY_REFRESH - 0x12) takes about 1.3 seconds (C1 to C2), which is faster compared to the original firmware (1.6 seconds). This is some quirk of the original firmware, which delays a bit the two data send sequences, but from my experiments it doesn't matter, so I left it as is in the driver code.

Following command DISPLAY_REFRESH (0x12), signal BUSY_N is low for 1.25 seconds (A1 to A2 and B1 to B2). This can be expected and the timings exactly matches that of the original firmware (i.e. waiting for de-asserting the BUSY_N signal).

@michapr You can use Saleae Logic to open the captures, if you need to peek into more details about the timings or the decoded SPI protocol (give me a shout if you need another format). I noticed some slight discrepancies in the intermediate re-initialisation of the display, I'll play with those at some point in order to bring my code as close as the original as possible in terms of SPI communication.


In conclusion, there is something else in play that contributes to the excessive energy consumption in the ATC firmware. I have no explanation at present. From my point of view, the SPI communication in the custom firmware closely matches that of the original firmware and this shouldn't be the contributing factor for the excessive energy consumption.

**Comment by michapr** [comment]
@znanev Thanks a lot for this details, I will check it (have the software in use)
What I see is that BUSY_N have in original FW some additional short peaks - looking at our code I thought it should be here too.
We are waiting multiple times for the BUSY_N state, right?

Maybe it make sense to set the CPU into longer wait state during the A1/A2 period? Every wakeup cost energy, maybe this also could help.

**Comment by znanev** [comment]
@michapr Indeed the original firmware probably awaits for BUSY_N after every command. In all fairness, the code I used as base for the EPD driver had waiting for BUSY_N on these places that are in the code now. At the time I didn't have good captures to compare like-for-like the commands, arguments and the BUSY_N and RST_N signals. Now that we have them, we can work through them to enhance the code to match as closely to the original SPI communication as possible. But this will take time...

**Comment by pvvx** [comment]
@michapr Where the use of Standby power consumption? (1.42..1.45 mA):
![image](https://user-images.githubusercontent.com/12629515/102036211-4af69580-3dd3-11eb-890b-a4b4321f181d.png)

**Comment by pvvx** [comment]
@michapr
![image](https://user-images.githubusercontent.com/12629515/102036620-5c8c6d00-3dd4-11eb-879d-cdc24201f26e.png)
Minimal deep-sleep-> test-> deep-sleep cycle for TLSR825x
![image](https://user-images.githubusercontent.com/12629515/102036796-d6245b00-3dd4-11eb-9a49-f33ba6825c5c.png)
(500uA / 200us grid step)
Example: Recharge deep-sleep (Timer + Retention 16 KiB SRAM):
(1.82 mA * 0.5 ms + 2.7 mA * 0.9 ms) / 1.4 ms

**Comment by pvvx** [comment]
cpu_sleep_wakeup(DEEPSLEEP_MODE_RET_SRAM_LOW16K , PM_WAKEUP_TIMER, clock_time() + x*CLOCK_16M_SYS_TIMER_CLK_1S) 
The maximum cpu_sleep_wakeup (DEEPSLEEP_MODE_RET_SRAM_LOW16K, PM_WAKEUP_TIMER, clock_time () + 8 * CLOCK_16M_SYS_TIMER_CLK_1S) in Telink_825X_SDK is only up to 8 seconds. (8 inclusive)
If more than 8 seconds, clock_time () counter and other timers counters will be invalid.

**Comment by pvvx** [comment]
_PM_WAKEUP_PAD wake-up source comes from GPIO module, all GPIOs except 4 MSPI pins
(PAx / PBx / PCx / PDx) high and low levels have wake-up function._
cpu_set_gpio_wakeup (IO_BUSY_N,Level_High, 1); and goto deep-sleep.

**Comment by pvvx** [comment]
> BLE is always broadcasting its name, in adidtion to the name it is "Advertising" the sensor data. It makes not difference if Advertising is enabled or not. Ble is sending every 3 seconds or so an name and advertising package even on stock firmware

Maximum - 4 sec. With a 3 second interval, there are already a lot of drops on the receiving side. This has most of the external software on different BLE devices.

**Comment by pvvx** [comment]
atc1441 > I wouldnt call myself a guru on that as well, my way would also be try and error, need to get my hands on a low power meassuring method!
GY-169 INA169, INA199 + oscilloscope. 
Shunt - ~5..100 Ohm

**Comment by michapr** [comment]
@pvvx 
> @michapr Where the use of Standby power consumption? (1.42..1.45 mA):

take the y-value/10 - you will get mA (is witten below the graphics ;) ) - and there is noise above the signal, I know.
I use a shunt 10 Ohm, so the voltage is low, and my oscilloscope will show noise without additional external filter.
It was not made for a "measurement", but only to get an idea about the energy consumption of the device...

Agree with ADC on/off, I have used the original FW only and tried to optimize the main procedures only, but less inside.
As I said, this is not "my CPU" - I use more the Atmel/ESP series and the nRF 5x series for BLE :)

I hope to find time to read more in the TLSR documentation - but it is not easy to make "experiments" on a ready device - better would be to have a breakout board like you use....

**Comment by pvvx** [comment]
> take the y-value/10 - you will get mA (is witten below the graphics ;) 
on your graph 3.3mA. ADS off = 3.3-0.4 = 2.9.

**Comment by pvvx** [comment]
ATC rf_set_power_level_index (RF_POWER_P3p01dBm);  TX = < 8 mA

**Comment by znanev** [comment]
@pvvx Thanks for the heads up about the interrupt waiting function. This is probably the way the original firmware is awaiting the BUSY_N de-assertion.

@michapr Can I please ask you to pull the code from branch [test-pin-wakeup](https://github.com/znanev/ATC_MiThermometer/compare/master...znanev:test-pin-wakeup), build it and test it on your device? Since I got a negative WAF (Wife Approval Factor) last week, I had to de-solder the debug wires of my device and re-assemble it back and clean-up my workspace :) So I have no means to upload firmware via serial if the last binary turns bad..

**Comment by znanev** [comment]
> Checking the original XiaomiLYWSD03MMC showed that it was not possible to use it. Large errors in temperature and humidity.

@pvvx What sensor did you compare that to? From what I can tell, all 5 Xiaomi sensors (2 x LYWSD02, 2 x MHO-C401 and one MHO-C303) in the same room agree on their values, so at least they are consistent and a simple factor could be used to compensate for temperature / humidity value, if needed.

**Comment by pvvx** [comment]
@znanev 
The wife spoke about the instability of the data readings. The effect is that the temperature changes and the reading freezes indefinitely when it drops. Some external conditions have not changed yet. It looks like the effect of the sensor falling off. Installed in an unheated room - unstable temperature from +10 to + 22C, if the door is open. This is the first thing.
But there is no precision itself. It is possible that at 20 + -5C it shows correctly, but in other ranges it is not.
I have almost all types of temperature and humidity sensors.
PS: I am writing via google translator - there may be incorrect descriptions.

**Comment by pvvx** [comment]
SHT85 - the best cheap on the market - does not change reading when polled. It does not heat up when active. The rest of the sensors depend on the number of polls and the current consumption during active measurement. Reaches +3 C.
By humidity. Condensation occurs in many sensors when the temperature changes and they give incorrect readings until they get hot. Applies to BME / BMP 280 - iron case with hole :) When turned on, it gives different temperature readings as it warms up. Up to +2 C.

**Comment by michapr** [comment]
@znanev 
>@michapr Can I please ask you to pull the code from branch test-pin-wakeup, build it and test it on your device? 

I have tested - energy usage is much better - but display does not work more ;)

Display is black with some white elements, they are changing when temperature will be changed, so something happen at display.
I could flash back via OTA - so you can try yourself without risk :)

**Comment by pvvx** [comment]
https://github.com/znanev/ATC_MiThermometer/blob/master/ATC_Thermometer/app.c#L85
overflow: 
5*60000*CLOCK_SYS_CLOCK_1MS = 5*60000*24000000/1000 = 7200000000 = 0x1 AD27 4800

**Comment by znanev** [comment]
> https://github.com/znanev/ATC_MiThermometer/blob/master/ATC_Thermometer/app.c#L85
> overflow:
> 5 * 60000 * CLOCK_SYS_CLOCK_1MS = 5 * 60000 * 24000000 / 1000 = 7200000000 = 0x1 AD27 4800
> 
> * Hundreds of warnings if gcc -pedantic

Indeed it overflows, I noticed it, but haven't have the time nor incentive to change it yet, as it's "inherited" from atc1441's codebase ;) 

My immediate goal is to make the EPD work like in the original firmware. Optimisations will follow after that, for sure :)

**Comment by michapr** [comment]
@znanev I agree, here in this task we should try work at the EPD at first ;)

@pvvx Some other optimizations (like this overflow) are already done in personal codebases - anyway thank you for reminder ;)

**Comment by znanev** [comment]
> @znanev
> 
> > @michapr Can I please ask you to pull the code from branch test-pin-wakeup, build it and test it on your device?
> 
> I have tested - energy usage is much better - but display does not work more ;)
> 
> Display is black with some white elements, they are changing when temperature will be changed, so something happen at display.
> I could flash back via OTA - so you can try yourself without risk :)

So it seems that the effect is the same as in #8. In that (bad) release, the only change was the sleep mode - instead of waiting in idle state:

```c
cpu_stall_wakeup_by_timer0(microseconds * CLOCK_SYS_CLOCK_1US);

```
the CPU was waiting in deep sleep state:
 
```c
cpu_sleep_wakeup(DEEPSLEEP_MODE_RET_SRAM_LOW32K, PM_WAKEUP_TIMER, clock_time() + microseconds * CLOCK_SYS_CLOCK_1US);
```

So from functional perspective, it appears that now

```c
cpu_set_gpio_wakeup(IO_BUSY_N, Level_High, 1);
```

behaves the same as the deep sleep waiting.

It should be something related to recovering from deep sleep, but I have no idea at this point. @pvvx - do you have any clue about this?

**Comment by znanev** [comment]
> But there is no precision itself. It is possible that at 20 + -5C it shows correctly, but in other ranges it is not.

This might be the key! All my sensors so far have been on room temperature, from 18C to 30C and their readings have generally been in the ballpark of a simple spirit thermometer. But I haven't tested them (yet) with extreme temperatures - like close to freezing point or more than 30C. So far I'm fine with my sensors, can't complain.
From what I know, MHO-C401 uses Swiss made sensors - Sensirion SHT30-DIS, so these in my opinion are pretty accurate.

> PS: I am writing via google translator - there may be incorrect descriptions.

No worries, I understand everything that you write so far. If there is something that I can't understand, I'll surely ask you to rephrase ;)

**Comment by znanev** [comment]
> This is the reason for "global warming". :)
> "global warming" with the introduction of electronic sensors:
> https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/Global_Temperature_And_Forces.svg/300px-Global_Temperature_And_Forces.svg.png

@pvvx Nice catch, hehe 😄  So even just measuring the temperature with those sensors, warming them up by 2 degrees, contributes to the global warming too 😛

**Comment by pvvx** [comment]
Doesn't work in the refrigerator. Shows 'L'

**Comment by pvvx** [comment]
https://github.com/znanev/ATC_MiThermometer/blob/master/ATC_Thermometer/flash.c#L12
```
	flash_write_page(0x78000, 1, (unsigned char *)&read);
```

**Comment by znanev** [comment]
> Doesn't work in the refrigerator. Shows 'L'

L probably stand for **L**ow. Fair enough, it seems that those sensors are suitable for room temperatures only, not outside or inside a fridge / freezer 😄

**Comment by znanev** [comment]
> https://github.com/znanev/ATC_MiThermometer/blob/master/ATC_Thermometer/flash.c#L12
> 
> ```
> 	flash_write_page(0x78000, 1, (unsigned char *)&read);
> ```

I haven't touched this code, yet. What do you want to point about this function call?

**Comment by pvvx** [comment]
LCD work at -20C, which is strange

**Comment by znanev** [comment]
EPD however won't work below 0C - its fluids will freeze!

**Comment by znanev** [comment]
Guys, do you mind starting a "discussion" here:

https://github.com/znanev/ATC_MiThermometer/discussions

It will probably be easier to catch up there.

**Comment by znanev** [comment]
> ```
> flash_write_page(0x78000, 1, read);
> ```
> 
> = flash_write_page(0x78000, 1, 0)
> flash_write_page declared as:
> 
> ```
> void flash_write_page(unsigned long addr, unsigned long len, unsigned char *buf)
> ```

It will be better to hint the original code maintainer - @atc1441 for that. I haven't touched / used that functionality yet.

**Comment by znanev** [comment]
I have a theory why changing the idle wait for deep sleep when waiting for BUSY_N to be de-asserted bugs the display.

Using idle wait doesn't allow the CPU to go into sleep, so the code on start-up - in function `user_init_normal`, which in turn invokes function `epd_init(&c, 1)`, is executed without interruption.

So when changing the idle wait for deep sleep, the first wakeup (and in fact all subsequent) of the CPU following the wait for BUSY_N goes into function `user_init_deepRetn`, which messes up the code in `epd_init` function. What happens is, only the first sequence from the [`redraw`](https://github.com/znanev/ATC_MiThermometer/blob/e0069a286140b7af7a8d964cbf7d4b74eda49924/ATC_Thermometer/epd.c#L247) condition is executed:

https://github.com/znanev/ATC_MiThermometer/blob/e0069a286140b7af7a8d964cbf7d4b74eda49924/ATC_Thermometer/epd.c#L256

This sequence sends initialisation waveforms for the black segments. After it is transmitted, the CPU waits for BUSY_N to become High again. And in the case of deep sleep, the `redraw` condition is false and the seconds sequence, initialising the waveforms for the white segments, won't be executed:

https://github.com/znanev/ATC_MiThermometer/blob/e0069a286140b7af7a8d964cbf7d4b74eda49924/ATC_Thermometer/epd.c#L271

Thinking forward, it appears that EPD driver should be completely rewritten as a state machine. Because after every wait for BUSY_N, resuming will begin the execution of the program from the beginning (in function `user_init_deepRetn` in app.c).

Until recently, my other goal was to keep the structure of the EPD driver as close to the "original" C++ code as possible:

https://github.com/znanev/MHO-C401

All was fine while I used the EPD alone, connected to another MCU - ESP8266 in that instance. But minimum power consumption was never in my requirements list so to say. So going forward, I think that the following should happen:

1. Complete break from the original EPD driver (i.e. compatibility not to be considered when doing modifications).
2. Think about how to re-implement sending each EPD command sequence as a state machine (because after every wake up following wait for BUSY_N, the program execution will in fact start from the beginning).
3. Optimise some of the sequences, bringing them in line with those from the original firmware (using the new logic analyser captures). 

I see myself working on 1 and 3 in the near future. Any volunteers for 2? 😉

**Comment by pvvx** [comment]
StartUp Original Xiaomi LYWSD03MMC
![image](https://user-images.githubusercontent.com/12629515/102118997-846de600-3e51-11eb-8607-eccb82b92dca.png)

**Comment by pvvx** [comment]
Work cycle (not connected):
![image](https://user-images.githubusercontent.com/12629515/102119761-aae05100-3e52-11eb-9dd8-5bf803bcb460.png)
Interval 1690.4 ms
average consumption 19.22 uA (3.3V)

**Comment by znanev** [comment]
> Work cycle (not connected):
> ![image](https://user-images.githubusercontent.com/12629515/102119761-aae05100-3e52-11eb-9dd8-5bf803bcb460.png)
> Interval 1690.4 ms
> average consumption 19.22 uA (3.3V)

@pvvx Is this for MHO-C401? If not, we are comparing oranges and apples, because LYWSD03MMC uses LCD, whereas MHO-C401 used EPD.

**Comment by pvvx** [comment]
> @pvvx Is this for MHO-C401? If not, we are comparing oranges and apples, because LYWSD03MMC uses LCD, whereas MHO-C401 used EPD.

I don't think there will be much difference.

**Comment by znanev** [comment]
> I don't think there will be much difference.

I may be wrong, but charging the charge pump of the EPD maybe consumes more energy compared to energising the LCD segments... But as you say, the difference might not be significant

**Comment by michapr** [comment]
@znanev 
>So when changing the idle wait for deep sleep, the first wakeup (and in fact all subsequent) of the CPU following the wait for BUSY_N goes into function user_init_deepRetn, which messes up the code in epd_init function. What happens is, only the first sequence from the redraw condition is executed:

Maybe we could use a flag to detect - we are within the send_sequence() function or not? And depending on this execute the needed functions in epd_init() ?

**Comment by pvvx** [comment]
The following charts alternate:
![image](https://user-images.githubusercontent.com/12629515/102125929-71601380-3e5b-11eb-8e51-93f39716347d.png)
Reading temperature and humidity sensor (?):
![image](https://user-images.githubusercontent.com/12629515/102125942-758c3100-3e5b-11eb-9d1b-658244617d0a.png)
To clarify what causes the 600 μA consumption, we need to solder the I2C pin ...

**Comment by pvvx** [comment]
The connection to "nRF Connect" is terrible. In half an hour, it will drain the battery ...
It needs to be redone.
![image](https://user-images.githubusercontent.com/12629515/102128882-8e96e100-3e5f-11eb-9829-cae76aab02f3.png)
![image](https://user-images.githubusercontent.com/12629515/102128896-93f42b80-3e5f-11eb-8b59-13c24018309f.png)

**Comment by michapr** [comment]
@pvvx 
>The connection to "nRF Connect" is terrible.

Yes, the main goal of the ATC firmware for MMC3 is to get the values without connection, because of battery drain.
So all values will be send in advertising package.
I have seen the same some time ago - but because do not need the connection, it was not important for me in his time.

But here - in this issue we try to solve at first the EPD integration...
Maybe you can have a look at the epd.c and give us some hints and help?

All the other issues would be very helpful to add in discussion - or in new issues, it will be better to sort it out then...
All what you say is very helpful to optimize the sensor - but this are different issues.we cannot work on all issues at same time :)
(BTW: I have all noted,... but... )

Thank you for your work and help!! 👍

**Comment by znanev** [comment]
I propose to stick to the EPD-related issues in this fork. The original codebase is something that I'm not very interested to dig into at present. It will be better if the original code is optimised first, later on we can merge the relevant parts here.

**Comment by pvvx** [comment]
This is an analogue of the ATC_Thermometer firmware. 
![image](https://user-images.githubusercontent.com/12629515/102131670-9b1d3880-3e63-11eb-8f1d-91864f966891.png)
There is still work to be done before EPD.

**Comment by michapr** [comment]
@znanev 
If I understand right we do not need all the epd_init()   within the send_sequence(), right?
I just excluded it using a RAM bool flag (epd_active true/false - set in the send_sequence() ) but no changes - display is black.
What I understood wrong?

What I have seen, adding a sleep_ms(100) at the end (after last cpu_set_gpio_wakeup()  ) - display will show data (with bad contrast and some old grey segments,...) - why this?

**Comment by michapr** [comment]
@pvvx 
>There is still work to be done before EPD.

YES - you are right !!! 
But could you please report this on the main repo at ATC? With new issues?
We are there also active and working there too.

But here in the **forked repo** we try to optimize the EPD driver only ;)

**Comment by pvvx** [comment]
> I propose to stick to the EPD-related issues in this fork. The original codebase is something that I'm not very interested to dig into at present. It will be better if the original code is optimised first, later on we can merge the relevant parts here.

For EPD, you need to prepare a deep sleep state machine. If the EPD control codes are correct, then you need to rewrite them under the state machine.

**Comment by znanev** [comment]
> @znanev
> If I understand right we do not need all the epd_init() within the send_sequence(), right?
> I just excluded it using a RAM bool flag (epd_active true/false - set in the send_sequence() ) but no changes - display is black.
> What I understood wrong?
> 
> What I have seen, adding a sleep_ms(100) at the end (after last cpu_set_gpio_wakeup() ) - display will show data (with bad contrast and some old grey segments,...) - why this?

The issue is not only within the epd_init() function, but send_sequence() itself - after every BUSY_N wake up from deep sleep, the MCU will start the program from "scratch" - only "remembering" the annotated RAM variables. That's why a FSM (finite state machine) is needed - defining the transitions and sequences of the EPD commands. I'll think about some crude implementation later on.

**Comment by pvvx** [comment]
After init_ble() use the bls_pm_registerFuncBeforeSuspend() function
```
void bls_pm_registerFuncBeforeSuspend (suspend_handler_t func );
```

```
int app_suspend_enter ()
{
	if (stop_sleep)
	{
		bls_pm_setSuspendMask(SUSPEND_DISABLE);
		return 0;
	}
	return 1;
}

bls_pm_registerFuncBeforeSuspend( &app_suspend_enter );
```

**Comment by pvvx** [comment]
```
	bls_app_registerEventCallback(BLT_EV_FLAG_SUSPEND_ENTER, &func_suspend_enter);
	bls_app_registerEventCallback(BLT_EV_FLAG_SUSPEND_EXIT, &func_suspend_exit);
	bls_app_registerEventCallback(BLT_EV_FLAG_GPIO_EARLY_WAKEUP, &func_gpio_wakeup);
```
or 
```
	blc_hci_registerControllerEventHandler(event_handler);		//register event callback
	int event_handler(u32 h, u8 *p, int n) {
		if (h &HCI_FLAG_EVENT_BT_STD) {		//ble controller hci event
			u8 evtCode = h & 0xff;
			if(evtCode == HCI_EVT_DISCONNECTION_COMPLETE)  //connection terminate
			{
			}
			else if(evtCode == HCI_EVT_ENCRYPTION_CHANGE)
			{
			}
			....
		}
		if (h &HCI_FLAG_EVENT_TLK_MODULE) {
			switch((u8)(h&0xff))
			case BLT_EV_FLAG_GPIO_EARLY_WAKEUP:
				break;
			case BLT_EV_FLAG_SUSPEND_ENTER:
				break;
			case BLT_EV_FLAG_SUSPEND_EXIT:
				break;
			....	
		}
	}
```

**Comment by znanev** [comment]
> What I see is that BUSY_N have in original FW some additional short peaks - looking at our code I thought it should be here too.
> We are waiting multiple times for the BUSY_N state, right?

@michapr Just checked the original firmware captures, BUSY_N is asserted (pulled low by the EPD) following one of these 3 commands:

- 0x02 (POWER_OFF)
- 0x04 (POWER_ON)
- 0x12 (DISPLAY_REFRESH)

The **epd.c** driver awaits for BUSY_N signal to be pulled back high for those 3 commands above too. I see that I've put waiting for BUSY_N for these two commands too:

- 0x18 (DATA_START_TRANSMISSION_1)
- 0x1c (DATA_START_TRANSMISSION_2)

Which serves no purpose, I'll remove these two waits.

**Comment by pvvx** [comment]
```
wakeup_callback_t wakeup_cb; // typedef void (*wakeup_callback_t)(void);

...
wakeup_cb =  read_sensor_cb;
...

_attribute_ram_code_ void read_sensor_cb(void) {
	wakeup_cb = NULL;
	....
}



_attribute_ram_code_ void blt_pm_proc(void) {
	if(ota_is_working){
		bls_pm_setSuspendMask(SUSPEND_DISABLE);
		bls_pm_setManualLatency(0);
	}else{
		if (wakeup_cb) {
			if(((bls_pm_getSystemWakeupTick() - clock_time())) > 30 * CLOCK_16M_SYS_TIMER_CLK_1MS) {
				cpu_set_gpio_wakeup(GPIO_WAKEUP_SENS, Level_High, 1);  // pad high wakeup deepsleep
				bls_pm_setWakeupSource(PM_WAKEUP_PAD);  // gpio pad wakeup suspend/deepsleep
			} else ...
		}
		bls_pm_setSuspendMask (SUSPEND_ADV | DEEPSLEEP_RETENTION_ADV | SUSPEND_CONN | DEEPSLEEP_RETENTION_CONN);
	}
}


attribute_ram_code_ void user_init_deepRetn(void){ // after sleep this will get executed
	blc_ll_initBasicMCU();
	rf_set_power_level_index(RF_POWER_P3p01dBm);
	blc_ll_recoverDeepRetention();

	if(pm_is_deepPadWakeup()) {
		if (wakeup_cb)	wakeup_cb();
	}
}
```
![image](https://user-images.githubusercontent.com/12629515/102205360-24705180-3edc-11eb-80e9-e0483e64ec65.png)

**Comment by znanev** [comment]
@pvvx I'm not sure I can follow up your thought process here. Can you please expand your examples with some explanations?

**Comment by pvvx** [comment]
In the main loop (main_loop), assign calback (). At the end of the main_loop() , blt_pm_proc () is called. It will set the modes for the sleep mode.
Then the MCU falls asleep and wakes up on the event specified: cpu_set_gpio_wakeup (GPIO_WAKEUP_SENS, Level_High, 1); // deep sleep for deep awakening.
On wake-up (in main.c) after deep-sleep, the user_init_deepRetn() routine is executed.
In it you can see that the activation came from the GPIO with the pm_is_deepPadWakeup (). 
... and execute calback(). 

This is probably the simplest example... Next, you need to enter your codes EPD.

**Comment by znanev** [comment]
@pvvx Thank you for your thorough explanation of the process above. So it seems that Telink's SDK has already implemented the hardest part - provide the plumbing for a state machine.

So, this brings the horizon of energy-efficient EPD driver rewrite a bit closer :)

I'll start with some code clean-up first, until I wrap my head around the implementation details of the state machine driver :)

**Comment by pvvx** [comment]
The SDK has "int event_handler (u32 h, u8 * para, int n)" for all events. For everyone events.

**Comment by znanev** [comment]
I wish we had the English version of the BLE_SDK_Developer_Handbook.pdf document. I can't make sense from most of the Google translated sentences... There are quite nice examples in there, along with explanations how to use different BLE-related functions. But my Chinese is non-existent, so I'll rely on somebody else to decipher that ;)

**Comment by pvvx** [comment]
http://wiki.telink-semi.cn/wiki/chip-series/TLSR825x-Series/
BLE Generic  TLSR8258  User Guide  SDK V4.0.0  Development Handbook
http://wiki.telink-semi.cn/doc/an/AN_20050601-E_Telink%20Kite%20Multiple%20Connection%20BLE%20SDK%20Developer%20Handbook.pdf

http://wiki.telink-semi.cn/tools_and_sdk/Driver/doc/kite/html/index.html

**Comment by pvvx** [comment]
There are problems with the SHTC3 sensor. The solution is described here
https://github.com/atc1441/ATC_MiThermometer/issues/134

instead of wake-up via GPIO, setting a wakeup event by timer is used
```
bls_pm_registerAppWakeupLowPowerCb(WakeupLowPowerCb);
bls_pm_setAppWakeupLowPower(clock_time() + x * CLOCK_16M_SYS_TIMER_CLK_1MS, 1);
```

**Comment by pvvx** [comment]
https://pvvx.github.io/MHO_C401/power_original.html

**Comment by pvvx** [comment]
![image](https://user-images.githubusercontent.com/12629515/104600095-9ed1f400-5689-11eb-824f-40b792995caa.png)
![image](https://user-images.githubusercontent.com/12629515/104600102-a1cce480-5689-11eb-9b3b-3dbf03e4f434.png)

**Comment by michapr** [comment]
@pvvx  I have updated the C401 some time ago with your code - it is working fine.
additional have added an EPD update only when value was changed - this save much energy.

For other room have added the option to change display only if temp. change >0.1 degree and hum. change >1% - also a possible option while we cannot save energy while updating ;)

**Comment by pvvx** [comment]
The original has a minimum consumption of 52 μA. Alternative firmware under the same conditions - 23 μA.
Less makes no sense...

**Comment by michapr** [comment]
About the 52μA for original FW I'm not sure... in any case.
I have 3 such sensors and all of them with different battery types had after less than 3 months about an empty battery (if using the BLE connection every 10 minutes).
After switching to advertising mode (using this repo with optimizing EPD write only if value is changing) it was much better.

Now - with your repo and adapted EPD it is much better. 
23μA - is it with the current EPD procedure? It take about 1.4seconds - and could not be optimized with sleeping process.

EDIT: I just have seen that you have updated the epd.c - with state.
I was working on it also this week - nice, can use now your repo !

Thank you!

**Comment by pvvx** [comment]
> 23μA - is it with the current EPD procedure? It take about 1.4seconds - and could not be optimized with sleeping process.

Version 12, default settings.
The sensor is on the table. In 140 seconds, the digits for output have changed 2 times. E-Inc update only on information change for it (0.1C, 1%).
Average 18.8 uA.
![image](https://user-images.githubusercontent.com/12629515/104686531-72ac8680-570e-11eb-815e-94880773b143.png)

The minimum time before the image change is set.
[TelinkMiFlasher.html](https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html)
Default settings:
Min. Display interval 10 (9.95) sec. (The period is preferably a multiple of the Advertising interval - 0.5..2%.)
Sensor polling time 20 seconds.
Advertising interval 2.5 seconds.

CR2032 > 100 mAh. 100/0.02 = 5000 hours

**Comment by znanev** [comment]
@atc1441 @michapr @pvvx Guys,

I'm closing this issue and will later archive this repo, as the active development now happens in Victor's repository :)

Thank you all for the brainstorming - it was a pleasure for me to discuss these matters with you.

I'm not saying farewell - far from it! Surely I'll be around lurking in your repos (whichever is currently active, that is).
